encoding/json/json_test.go

```diff
@@ -17,70 +19,101 @@ type subsetObject struct {
-		var decoded subsetObject
-		err := decoder.Decode(&decoded)
-		require.Error(t, err)
-		assert.Contains(t, err.Error(), "no matching struct field found when decoding stream map with key name")
-	})
-
-	t.Run("lenient decode subset, pass", func(t *testing.T) {
-		// strict decode test
-		decoder := NewLenientDecoder(bytes.NewReader(encodedOb))
-		var decoded subsetObject
-		err := decoder.Decode(&decoded)
-		require.NoError(t, err)
-		assert.Equal(t, obj.subsetObject, decoded)
-	})
-
-	t.Run("original encode map key as string", func(t *testing.T) {
-		intMap := map[int]string{
-			0: "int key",
-		}
-		data := string(Encode(intMap))
-		assert.NotContains(t, data, "\"0\":")
-	})
-
-	t.Run("strict encode map key as string", func(t *testing.T) {
-		intMap := map[int]string{
-			0: "int key",
-		}
-		data := string(EncodeStrict(intMap))
-		assert.NotContains(t, data, "0:")
-	})
-
-	t.Run("strict encode map interface key as string", func(t *testing.T) {
-		t.Skip("There is a bug in go-codec with MapKeyAsString = true and Canonical = true")
-		intMap := map[interface{}]interface{}{
-			0: "int key",
-		}
-		data := string(EncodeStrict(intMap))
-		assert.NotContains(t, data, "0:")
-	})
+	encodedOb = Encode(obj)
+}
+
+func TestBasicEncodeDecode(t *testing.T) {
+	// basic encode/decode test.
+	var decoded object
+	err := Decode(encodedOb, &decoded)
+	require.NoError(t, err)
+	assert.Equal(t, obj, decoded)
+}
+
+func TestDecode(t *testing.T) {
+	decoder := NewDecoder(bytes.NewReader(encodedOb))
+	var decoded object
+	err := decoder.Decode(&decoded)
+	require.NoError(t, err)
+	assert.Equal(t, obj, decoded)
+}
+
+func TestSubsetDecode(t *testing.T) {
+	decoder := NewDecoder(bytes.NewReader(encodedOb))
+	var decoded subsetObject
+	err := decoder.Decode(&decoded)
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), "no matching struct field found when decoding stream map with key name")
+}
+
+func TestLenientDecode(t *testing.T) {
+	decoder := NewLenientDecoder(bytes.NewReader(encodedOb))
+	var decoded subsetObject
+	err := decoder.Decode(&decoded)
+	require.NoError(t, err)
+	assert.Equal(t, obj.subsetObject, decoded)
+}
+
+func TestEncodeMaapKeyAsString(t *testing.T) {
+	intMap := map[int]string{
+		0: "int key",
+	}
+	data := string(Encode(intMap))
+	assert.NotContains(t, data, `"0"`)
+}
+
+func TestStrictEncodeMapIntKeyAsString(t *testing.T) {
+	intMap := map[int]string{
+		0: "int key",
+	}
+	data := string(EncodeStrict(intMap))
+	assert.NotContains(t, data, "0:")
+}
+
+func TestStrictEncodeMapInterfaceKeyAsString(t *testing.T) {
+	intMap := map[interface{}]interface{}{
+		0: "int key",
+	}
+	data := string(EncodeStrict(intMap))
+	assert.Contains(t, data, `"0"`)
+}
+
+func TestStructKeyEncode(t *testing.T) {
+	type KeyStruct struct {
+		Key1 string `json:"key1"`
+		Key2 string `json:"key2"`
+	}
+	type TestStruct struct {
+		Complex map[KeyStruct]string `json:"complex"`
+	}
+
+	data := TestStruct{
+		Complex: map[KeyStruct]string{
+			KeyStruct{
```

Comment:
Linter says to remove `KeyStruct`:

```suggestion
			{
```